library/Fission/Web/IPFS/DAG.hs

```diff
@@ -43,5 +43,9 @@ put (Entity userId _) (Serialized rawData) =
           Web.Err.throw err
 
         Right pinnedCID -> do
-          _ <- User.CID.createX userId [newCID]
+          newCID
+            |> return
+            |> User.CID.createX userId
```

Comment:
It seems like `User.CID.Creator` should have to be in the constraints in order to call this?

It seems like some places, you include them in the constraints (`Fission.Web.Heroku.Deprovision`) and other cases you don't (here)

Reply:
Oh wait I think I get the difference in where they're called. This one, the Monad is `MonadDB` for running the txs. in `Heroku.Deprovision`, the constraints are used to construct the txs and then they're actually run in `MonadDB` in `destroy`.

Still seems weird that you can call `createX` without being instead a `User.CID.Creator` monad... Or are we inside one, and it's just implicit?

Reply:
@dholms Ah, yes, sorry I had meant to add this to the write up. This is called "contsraint plucking".

# Plucking

Just like how `runRIO` in `app/Main.hs` takes some `RIO cfg a` action, and turns it into `IO ()`, here we're taking some `Transaction m a` and turning it into `m a`. If you look at the definition of `runDB`...

```haskell
class MonadDB m where
  runDB :: Transaction m a -> m a
```

...you see that it takes something more specific (a `Transaction`) and turns it into something less specific. Now look at one of the model typeclasses:

```haskell
class User.Retriever m where
  getById :: m (Maybe User)

instance User.Retriever (Transaction m) where
  getById uid = selectFirst [UserId ==. uid]
```

We're saying that this will work on _any_ `m` (i.e. helpful for testing functions without `runDB`). But the one that we use in production is specifically a `Transaction`, which conveniently matches the input of `runDB`.

The original plan was to have each handler specify which resources it had access to. This put us in one of two places:

1. Use `runDB` in each `instance`, so we unwrap the `Transaction` per call. This loses us the database transaction and query optimizer. Not great, but lets us specify a constraint on any function in a really straightforward way. This PR does not do this.

2. Move `runDB` outside of `instances` to... somewhere else, ideally as high up in the stack as possible. But that means that everything is running inside `Transaction m`... which is actually totally fine, but means that all of our instances now need to know how to run inside a `Transaction`. This is automatic for many, but not all of them. So instead of writing a bunch of custom instances and orphans, this PR "plucks" the constraint out by giving each handler DB access and running the transaction.

We can still write helpers that constrain helpers, but at the end of the day the application has access to the DB. It's a trade-off, but IMO feels like this is the balance to strike: we can break out sensitive code (via helpers), mock everything out in tests, get access to the automatic query optimizer, and everything runs inside DB transactions, but place it in a more general / looser context to avoid a bunch of boilerplate.

# Not a Monad

> Still seems weird that you can call createX without being instead a User.CID.Creator monad... Or are we inside one, and it's just implicit?

A quick note on terminology: not all constraints are monadic, but "the monad class" is a constraint. We can run these in a concrete monad (like "List" or "Maybe") that _also_ implements the constraint `User.CID.Creator`.

> Or are we inside one, and it's just implicit?

With `User.CID.Creator`, we're inside a monad (data type) that _knows how to make `UserCid` creation DB transactions_. To actually execute those, we need to hand them to `runDB`. There's actually nothing implicitly monadic about creating a DB transaction — my first draft didn't even have `Monad m =>` on it, but we use them so frequently in monadic contexts that it was annoying to have to add an additional constraint to every function.

So, in production code, `createX` makes a `Transaction (RIO cfg)` function that will later be interpreted and executed by `runDB`. In tests, we're going to have something like:

```haskell
instance MonadDB (TestMock a) where
  runDB :: Transaction (TestMock a)
  runDB (Transaction toRun) = toRun

instance User.CID.Retriever (TestMock a) where
  getById _ = return (Just fakeUser1)
```

Reply:
Nope! Just ignore me, I think that I have a way of constraining `MonadDB` to specify the various access constraints! Whuich is great since as previously stated I'm not super thrilled with the tradeoffs as they are but it seemed like a balance. I think we can have our cake and eat it too!

This is why PRs are helpful 😅

Reply:
Aaaaand done

Reply:
ok sweet yeah this makes sense. Love how much information we get about functions from the new way of doing it